opslevel/resource_opslevel_domain.go

```diff
@@ -97,75 +97,75 @@ func (r *DomainResource) Schema(ctx context.Context, req resource.SchemaRequest,
 }
 
 func (r *DomainResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
-	var data DomainResourceModel
+	var planModel DomainResourceModel
 
 	// Read Terraform plan data into the model
-	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &planModel)...)
 
 	if resp.Diagnostics.HasError() {
 		return
 	}
 
 	resource, err := r.client.CreateDomain(opslevel.DomainInput{
-		Description: opslevel.RefOf(data.Description.ValueString()),
-		Name:        opslevel.RefOf(data.Name.ValueString()),
-		Note:        opslevel.RefOf(data.Note.ValueString()),
-		OwnerId:     opslevel.NewID(data.Owner.ValueString()),
+		Description: planModel.Description.ValueStringPointer(),
+		Name:        opslevel.RefOf(planModel.Name.ValueString()),
+		Note:        planModel.Note.ValueStringPointer(),
+		OwnerId:     opslevel.NewID(planModel.Owner.ValueString()),
 	})
 	if err != nil {
 		resp.Diagnostics.AddError("opslevel client error", fmt.Sprintf("Unable to create domain, got error: %s", err))
 		return
 	}
-	createdDomainResourceModel, diags := NewDomainResourceModel(ctx, *resource)
+	createdDomainResourceModel, diags := NewDomainResourceModel(ctx, *resource, planModel)
 	resp.Diagnostics.Append(diags...)
 
 	tflog.Trace(ctx, "created a domain resource")
 	resp.Diagnostics.Append(resp.State.Set(ctx, &createdDomainResourceModel)...)
 }
 
 func (r *DomainResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
-	var data DomainResourceModel
+	var stateModel DomainResourceModel
 
 	// Read Terraform prior state data into the model
-	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+	resp.Diagnostics.Append(req.State.Get(ctx, &stateModel)...)
 
 	if resp.Diagnostics.HasError() {
 		return
 	}
 
-	resource, err := r.client.GetDomain(data.Id.ValueString())
+	resource, err := r.client.GetDomain(stateModel.Id.ValueString())
 	if err != nil {
 		resp.Diagnostics.AddError("opslevel client error", fmt.Sprintf("Unable to read domain, got error: %s", err))
 		return
 	}
-	readDomainResourceModel, diags := NewDomainResourceModel(ctx, *resource)
+	readDomainResourceModel, diags := NewDomainResourceModel(ctx, *resource, stateModel)
```

Comment:
So this weird to me.  You are logically changing how this function `NewDomainResourceModel ` works between Read and Create because you are passing in the plan OR the state.  Its not 100% clear to me but i would venture to guess that creates different outcomes and the function `NewDomainResourceModel` shouldn't have different outcomes.

Additionally you then have to create all these weird functions like 
```StringValueFromResourceAndModelField(domain.Note, givenModel.Note)```
and
```StringValueFromResourceAndModelField(string(domain.Owner.Id()), givenModel.Owner)```

Thats a HUGE code smell that you shouldn't have a reusuable function here because logically there are differents in how you build the model struct that gets saved into the state - but its because you are trying to preserve the common function ... but in the end are making things way more complicated then if you didn't have the function and just inlined it.

I'm worried we are seeing alot of bugs due to this and long term this is going to make the provider impossible to maintain.

Thoughts?